### PR TITLE
Implementing the new interfaces

### PR DESCRIPTION
**Pull request by diasdavid**
> WIP

ref: https://github.com/ipfs/interface-ipfs-core/pull/162

### CODE REVIEW

File: src/core/components/files.js [hunk 1]
```diff
@@ -86,23 +177,40 @@ module.exports = function files (self) {
...
-          if (err) {
-            return callback(err)
-          }
-          if (!files || !files.length) return callback(new Error('No such file'))
-          callback(null, toStream.source(files[files.length - 1].content))
-        })
+        p,
+        pull.concat(callback)
       )
     }),
```
Comment (diasdavid) [inline src/core/components/files.js]: @pgte the .cat for a large file (15Mb) is failing with:

image

Seems that some of the streams is giving up too early. Could you help me figuring this one out?

**Comment by diasdavid** [comment]
What is missing in this PR:

- [ ] Fix big files tests
- [ ] Finish migrating exchange files to pull-streams
- [ ] Update remaining tests that use add, cat and get

File: src/core/components/files.js (hunk 1 above)
Comment (diasdavid) [reply src/core/components/files.js]: Think I got it, I can't use concat to concat buffers https://github.com/pull-stream/pull-stream/blob/master/sinks/concat.js

**Comment by diasdavid** [comment]
Now the example is using pull-streams directly (though ipfs.files.add) but I'm being unsuccessful to add and fetch files (although tests are 100% passing). Something might have changed in the browser APIs that is messing with the data blobs.

File: examples/browser-browserify/src/index.js [hunk 2]
```diff
@@ -1,18 +1,13 @@
...
 // var IPFS = require('ipfs')
 
 const node = new IPFS({
   repo: String(Math.random() + Date.now())
 })
 
-node.on('ready', () => {
-  console.log('IPFS node is ready')
-})
+node.on('ready', () => console.log('IPFS node is ready'))
```
Comment (pgte) [inline examples/browser-browserify/src/index.js]: What about using `.once` instead of `.on´?

File: examples/browser-script-tag/index.html [hunk 3]
```diff
@@ -4,33 +4,17 @@
...
-    node.init(handleInit)
-
-    function handleInit (err) {
-      if (err) {
-        throw err
-      }
-
-      node.start(() => {
-        console.log('Online status: ', node.isOnline() ? 'online' : 'offline')
+    node.on('ready', () => {
```
Comment (pgte) [inline examples/browser-script-tag/index.html]: Use `.once` instead?

File: examples/browser-webpack/src/components/app.js [hunk 4]
```diff
@@ -25,9 +25,7 @@ class App extends React.Component {
     function create () {
       // Create the IPFS node instance
 
-      node = new IPFS({
-        repo: String(Math.random() + Date.now())
-      })
+      node = new IPFS({ repo: String(Math.random() + Date.now()) })
 
       node.on('ready', () => {
```
Comment (pgte) [inline examples/browser-webpack/src/components/app.js]: `once` would be nicer, I think.

File: examples/exchange-files-in-browser/public/js/app.js [hunk 5]
```diff
@@ -33,16 +34,21 @@ function start () {
...
-        peerInfo = id
-        updateView('ready', node)
-        setInterval(refreshPeerList, 1000)
-        $peers.innerHTML = '<h2>peers</h2><i>waiting for peers...</i>'
-      })
-    })
+    // EXAMPLE
+    // node = new self.Ipfs({ repo: 'ipfs-' + Math.random() })
+
+    node.on('start', () => node.id((err, id) => {
```
Comment (pgte) [inline examples/exchange-files-in-browser/public/js/app.js]: what about `.once`?

File: test/core/kad-dht.node.js [hunk 6]
```diff
@@ -6,11 +6,10 @@ const chai = require('chai')
 const dirtyChai = require('dirty-chai')
 const expect = chai.expect
 chai.use(dirtyChai)
-const bl = require('bl')
 const parallel = require('async/parallel')
 const IPFSFactory = require('../utils/ipfs-factory-instance')
 
-describe('verify that kad-dht is doing its thing', () => {
+describe.skip('verify that kad-dht is doing its thing', () => {
```
Comment (diasdavid) [inline test/core/kad-dht.node.js]: This tests needs to be unskiped and passing before merging

**Comment by diasdavid** [comment]
Discovered with this PR's that http, gateway and CLI tests were being skipped in CI since the Circuit Relay PR, enabling them found a bunch of bugs, fixing them now.

File: test/cli/daemon.js [hunk 7]
```diff
@@ -71,33 +71,37 @@ describe('daemon', () => {
 
   afterEach(() => clean(repoPath))
 
-  it(`don't crash if Addresses.Swarm is empty`, function (done) {
-    this.timeout(20000)
+  // TODO: test fails
+  it.skip('do not crash if Addresses.Swarm is empty', function (done) {
```
Comment (diasdavid) [inline test/cli/daemon.js]: ref: https://github.com/ipfs/js-ipfs/pull/1086#issuecomment-345277543

File: test/cli/daemon.js [hunk 8]
```diff
@@ -71,33 +71,37 @@ describe('daemon', () => {
...
-      done(err)
-    })
+    }).catch((err) => done(err))
   })
 
-  it(`should handle SIGINT gracefully`, function (done) {
-    this.timeout(20000)
-    killSig('SIGINT').then(() => {
+  // TODO: test fails
+  it.skip('should handle SIGINT gracefully', function (done) {
```
Comment (diasdavid) [inline test/cli/daemon.js]: ref: https://github.com/ipfs/js-ipfs/pull/1086#issuecomment-345277543

File: test/cli/daemon.js [hunk 9]
```diff
@@ -71,33 +71,37 @@ describe('daemon', () => {
...
+    testSignal(ipfs, 'SIGINT').then(() => {
       checkLock(repoPath, done)
     }).catch(done)
   })
 
-  it(`should handle SIGTERM gracefully`, function (done) {
-    this.timeout(20000)
-    killSig('SIGTERM').then(() => {
+  // TODO: test fails
+  it.skip('should handle SIGTERM gracefully', function (done) {
```
Comment (diasdavid) [inline test/cli/daemon.js]: ref: https://github.com/ipfs/js-ipfs/pull/1086#issuecomment-345277543

**Comment by diasdavid** [comment]
# Update, I'm down to:

### Core `npm run test:unit:node:core`

image

### HTTP `npm run test:unit:node:http`

screen shot 2017-11-17 at 16 27 39

### CLI `npm run test:unit:node:cli`

screen shot 2017-11-17 at 16 36 59

The HTTP and CLI failures are related, fixing one will fix the other. Both of them throw the following error:

screen shot 2017-11-17 at 16 35 48

**Comment by dryajov** [comment]
@diasdavid 

> Discovered with this PR's that http, gateway and CLI tests were being skipped in CI since the Circuit Relay PR, enabling them found a bunch of bugs, fixing them now.

Were those ever enabled? I can't trace them to being enabled either in `package.json` nor travis/circle.

**Comment by diasdavid** [comment]
@dryajov yes, they were enabled through the node.js file that was present in the test folder (like it is now)

File: examples/exchange-files-in-browser/public/js/app.js [hunk 10]
```diff
@@ -70,41 +74,25 @@ function createFileBlob (data, multihash) {
...
-          const listItem = createFileBlob(buf, multihash)
-
-          $filesList.insertBefore(listItem, $filesList.firstChild)
-        })
-
-        file.content.resume()
+        console.log('Fetched file:', cid, file.content.length)
+        // calling createFileBlob makes the Chrome go "Oh Snap"
+        // const listItem = createFileBlob(file.content, cid)
+        // $filesList.insertBefore(listItem, $filesList.firstChild)
```
Comment (diasdavid) [inline examples/exchange-files-in-browser/public/js/app.js]: Ok, narrowed down to the createFileBlob func. Apparently it always makes the Chrome go "oh Snap" and in Firefox it creates a wrong file (3x the size)

@VictorBjelkholm thoughts on this one?

**Comment by diasdavid** [comment]
# Good news everyone!

I was successfully able to upload a file with more than 750Mb (didn't test higher) and was able to fetch it through the gateway (slowly but it went through).

File: src/http/api/resources/files.js [hunk 11]
```diff
@@ -87,57 +87,57 @@ exports.get = {
...
 
-          pack.finalize()
-        })
-      )
+          header.type = 'directory'
+          pack.entry(header, () => {
+            console.log('packed a directory')
+            cb()
+          })
+        }
```
Comment (diasdavid) [inline src/http/api/resources/files.js]: The final issue is here, for some reason, tar-stream blocks on the second dir that is added:

image

**Comment by diasdavid** [comment]
Fixed it 🌟 The issue was introduced when I adopted the getPullStream for the http-api (instead of the old with ReadableStreams) and although both implemented just fine, the end of a `toStream.souce(pull-stream).pipe(tarPackStream)` was telling the whole pack that the thing had finished already and so it would stop after the first file.

**Pull request merged**